tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+                                       is_top ?
+                                       gap_start_pos_y - constants.pipe_cap_height * 0.5f :
+                                       gap_start_pos_y + constants.gap_height + constants.pipe_cap_height * 0.5f
+        };
+
+        // Construct the cap
+        auto cap = geometry::blueprint_rectangle(registry, cap_size, constants.pipe_color, cap_pos,
+                                                 constants.pipe_outline_color);
+
+        // Set layers, cap should be in front of body
+        registry.assign<graphics::layer<4>>(cap);
+        registry.assign<graphics::layer<3>>(body);
+        registry.assign<entt::tag<"game_scene"_hs>>(cap);
+        registry.assign<entt::tag<"game_scene"_hs>>(body);
+        registry.assign<entt::tag<"dynamic"_hs>>(cap);
+        registry.assign<entt::tag<"dynamic"_hs>>(body);
+
+        // Construct a pipe with body and cap and return it
+        return {body, cap};
+    }
+
+    // Factory to create single column
+    void create_column(entt::registry &registry, float pos_x) noexcept {
+        // Create a fresh entity for a new column
+        auto entity_column = registry.create();
+
+        // Get a random gap start position Y, between pipes
+        float gap_start_pos_y = get_random_gap_start_pos(registry);
+
+        // Create pipes, is_top variable is false for bottom one
+        auto top_pipe = create_pipe(registry, true, pos_x, gap_start_pos_y);
+        auto bottom_pipe = create_pipe(registry, false, pos_x, gap_start_pos_y);
+
+        // Make a column from these two pipes and mark it as "column"
+        registry.assign<column>(entity_column, top_pipe, bottom_pipe);
+        registry.assign<entt::tag<"column"_hs>>(entity_column);
+        registry.assign<entt::tag<"game_scene"_hs>>(entity_column);
+        registry.assign<entt::tag<"dynamic"_hs>>(entity_column);
+    }
+
+    // Factory for creating a Flappy Bird columns
+    void create_columns(entt::registry &registry) noexcept {
+        // Retrieve constants
+        const auto constants = registry.ctx<flappy_bird_constants>();
+
+        // Spawn columns out of the screen, out of the canvas
+        const float column_pos_offset = constants.column_start_distance + constants.column_thickness * 2.0f;
+
+        // Create the columns
+        for (std::size_t i = 0; i < constants.column_count; ++i) {
+            // Horizontal position (X) increases for every column, keeping the distance
+            float pos_x = column_pos_offset + i * constants.column_distance;
+
+            create_column(registry, pos_x);
+        }
+    }
+
+    // Factory for creating a Flappy Bird background
+    void create_background(entt::registry &registry) noexcept {
+        // Retrieve constants
+        const auto[canvas_width, canvas_height] = registry.ctx<graphics::canvas_2d>().canvas.size;
+        const auto constants = registry.ctx<flappy_bird_constants>();
+
+        // Create Sky
```

Comment:
Well, I always think that's better to separate functions, because one function one functionality, but here it's as you wish you think it's more readable let's keep it like this

Reply:
yeah I think it's fine since they are all background and working the same way